## ARGO Pipelines Processed WXS Datasets
**ARGO WorkflowName**|**analysisType**|**DataCategory**|**tumourNormalDesignation**|**FileType**|**OpenBucketURL**
:-----:|:-----:|:-----:|:-----:|:-----:|:-----:
DNA Seq Alignment|sequencing\_alignment|Sequencing Reads|tumour|CRAM|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WXS/sequencing\_alignment/tumour/TCRB-CA.DO262483.SA622771.wxs.20211126.aln.cram
DNA Seq Alignment|sequencing\_alignment|Sequencing Reads|tumour|CRAI|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WXS/sequencing\_alignment/tumour/TCRB-CA.DO262483.SA622771.wxs.20211126.aln.cram.crai
DNA Seq Alignment|sequencing\_alignment|Sequencing Reads|normal|CRAM|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WXS/sequencing\_alignment/normal/TCRB-CA.DO262483.SA622770.wxs.20211126.aln.cram
DNA Seq Alignment|sequencing\_alignment|Sequencing Reads|normal|CRAI|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WXS/sequencing\_alignment/normal/TCRB-CA.DO262483.SA622770.wxs.20211126.aln.cram.crai
DNA Seq Alignment|sequencing\_alignment|Metadata|tumour|JSON|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WXS/sequencing\_alignment/tumour/d5f2bbc1-8fed-4b8b-b2bb-c18fedbb8be4.sequencing\_alignment.json
DNA Seq Alignment|sequencing\_alignment|Metadata|normal|JSON|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WXS/sequencing\_alignment/normal/9f847b3c-6c98-427c-847b-3c6c98527c65.sequencing\_alignment.json
DNA Seq Alignment|qc\_metrics|Quality Control Metrics|tumour|TGZ|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WXS/qc\_metrics/tumour/TCRB-CA.DO262483.SA622771.wxs.20211126.aln.cram.qc\_metrics.tgz
DNA Seq Alignment|qc\_metrics|Quality Control Metrics|tumour|TGZ|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WXS/qc\_metrics/tumour/TCRB-CA.DO262483.SA622771.wxs.20211126.aln.cram.oxog\_metrics.tgz
DNA Seq Alignment|qc\_metrics|Quality Control Metrics|tumour|TGZ|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WXS/qc\_metrics/tumour/TCRB-CA.DO262483.SA622771.wxs.20211126.aln.cram.duplicates\_metrics.tgz
DNA Seq Alignment|qc\_metrics|Quality Control Metrics|tumour|TGZ|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WXS/qc\_metrics/tumour/TCRB-CA.DO262483.SA622771.wxs.20211126.TCRBOA7T-WXS.ubam\_qc\_metrics.tgz
DNA Seq Alignment|qc\_metrics|Quality Control Metrics|normal|TGZ|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WXS/qc\_metrics/normal/TCRB-CA.DO262483.SA622770.wxs.20211126.aln.cram.qc\_metrics.tgz
DNA Seq Alignment|qc\_metrics|Quality Control Metrics|normal|TGZ|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WXS/qc\_metrics/normal/TCRB-CA.DO262483.SA622770.wxs.20211126.aln.cram.oxog\_metrics.tgz
DNA Seq Alignment|qc\_metrics|Quality Control Metrics|normal|TGZ|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WXS/qc\_metrics/normal/TCRB-CA.DO262483.SA622770.wxs.20211126.aln.cram.duplicates\_metrics.tgz
DNA Seq Alignment|qc\_metrics|Quality Control Metrics|normal|TGZ|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WXS/qc\_metrics/normal/TCRB-CA.DO262483.SA622770.wxs.20211126.TCRBOA7N-WXS.ubam\_qc\_metrics.tgz
DNA Seq Alignment|qc\_metrics|Metadata|tumour|JSON|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WXS/qc\_metrics/tumour/abeb10ae-ab7f-410b-ab10-aeab7f310bf7.qc\_metrics.json
GATK Mutect2 Variant Calling|variant\_calling|Simple Nucleotide Variation|tumour|TBI|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WXS/variant\_calling/tumour/TCRB-CA.DO262483.SA622771.wxs.20211126.gatk-mutect2.somatic.snv.vcf.gz.tbi
GATK Mutect2 Variant Calling|variant\_calling|Simple Nucleotide Variation|tumour|TBI|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WXS/variant\_calling/tumour/TCRB-CA.DO262483.SA622771.wxs.20211126.gatk-mutect2.somatic.indel.vcf.gz.tbi
GATK Mutect2 Variant Calling|variant\_calling|Simple Nucleotide Variation|tumour|VCF|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WXS/variant\_calling/tumour/TCRB-CA.DO262483.SA622771.wxs.20211126.gatk-mutect2.somatic.snv.vcf.gz
GATK Mutect2 Variant Calling|variant\_calling|Simple Nucleotide Variation|tumour|VCF|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WXS/variant\_calling/tumour/TCRB-CA.DO262483.SA622771.wxs.20211126.gatk-mutect2.somatic.indel.vcf.gz
GATK Mutect2 Variant Calling|variant\_calling|Metadata|tumour|JSON|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WXS/variant\_calling/tumour/8c58e635-443c-4b05-98e6-35443ccb05a7.variant\_calling.json
GATK Mutect2 Variant Calling|variant\_calling|Metadata|tumour|JSON|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WXS/variant\_calling/tumour/7af5faf1-ceaf-46e0-b5fa-f1ceaf76e010.variant\_calling.json
GATK Mutect2 Variant Calling|qc\_metrics|Quality Control Metrics|tumour|TGZ|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WXS/qc\_metrics/tumour/TCRB-CA.DO262483.SA622771.wxs.20211126.gatk-mutect2.somatic.mutect\_filtering\_metrics.tgz
GATK Mutect2 Variant Calling|qc\_metrics|Quality Control Metrics|tumour|TGZ|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WXS/qc\_metrics/tumour/TCRB-CA.DO262483.SA622771.wxs.20211126.gatk-mutect2.somatic.mutect\_callable\_metrics.tgz
GATK Mutect2 Variant Calling|qc\_metrics|Quality Control Metrics|tumour|TGZ|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WXS/qc\_metrics/tumour/TCRB-CA.DO262483.SA622771.wxs.20211126.gatk-mutect2.somatic.contamination\_metrics.tgz
GATK Mutect2 Variant Calling|qc\_metrics|Quality Control Metrics|tumour|TGZ|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WXS/qc\_metrics/tumour/TCRB-CA.DO262483.SA622770.wxs.20211126.gatk-mutect2.somatic.contamination\_metrics.tgz
GATK Mutect2 Variant Calling|qc\_metrics|Metadata|tumour|JSON|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WXS/qc\_metrics/tumour/9030fb1b-3ea8-40b9-b0fb-1b3ea820b965.qc\_metrics.json
GATK Mutect2 Variant Calling|qc\_metrics|Metadata|normal|JSON|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WXS/qc\_metrics/normal/c21baaea-850b-4b84-9baa-ea850bcb84fa.qc\_metrics.json

