## ARGO Pipelines Processed WGS Datasets
**ARGO WorkflowName**|**analysisType**|**DataCategory**|**tumourNormalDesignation**|**FileType**|**OpenBucketURL**
:-----:|:-----:|:-----:|:-----:|:-----:|:-----:
DNA Seq Alignment|sequencing\_alignment|Sequencing Reads|tumour|CRAM|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WGS/sequencing\_alignment/tumour/TCRB-CA.DO262483.SA622771.wgs.20211117.aln.cram
DNA Seq Alignment|sequencing\_alignment|Sequencing Reads|tumour|CRAI|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WGS/sequencing\_alignment/tumour/TCRB-CA.DO262483.SA622771.wgs.20211117.aln.cram.crai
DNA Seq Alignment|sequencing\_alignment|Sequencing Reads|normal|CRAM|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WGS/sequencing\_alignment/normal/TCRB-CA.DO262483.SA622770.wgs.20211118.aln.cram
DNA Seq Alignment|sequencing\_alignment|Sequencing Reads|normal|CRAI|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WGS/sequencing\_alignment/normal/TCRB-CA.DO262483.SA622770.wgs.20211118.aln.cram.crai
DNA Seq Alignment|sequencing\_alignment|Metadata|tumour|JSON|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WGS/sequencing\_alignment/tumour/f31834bc-fc8b-4bec-9834-bcfc8babecf4.sequencing\_alignment.json
DNA Seq Alignment|sequencing\_alignment|Metadata|normal|JSON|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WGS/sequencing\_alignment/normal/e6207a9d-d134-486f-a07a-9dd134486f82.sequencing\_alignment.json
DNA Seq Alignment|qc\_metrics|Quality Control Metrics|tumour|TGZ|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WGS/qc\_metrics/tumour/TCRB-CA.DO262483.SA622771.wgs.20211117.aln.cram.qc\_metrics.tgz
DNA Seq Alignment|qc\_metrics|Quality Control Metrics|tumour|TGZ|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WGS/qc\_metrics/tumour/TCRB-CA.DO262483.SA622771.wgs.20211117.aln.cram.oxog\_metrics.tgz
DNA Seq Alignment|qc\_metrics|Quality Control Metrics|tumour|TGZ|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WGS/qc\_metrics/tumour/TCRB-CA.DO262483.SA622771.wgs.20211117.aln.cram.duplicates\_metrics.tgz
DNA Seq Alignment|qc\_metrics|Quality Control Metrics|tumour|TGZ|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WGS/qc\_metrics/tumour/TCRB-CA.DO262483.SA622771.wgs.20211117.TCRBOA7T-WGS-L4.ubam\_qc\_metrics.tgz
DNA Seq Alignment|qc\_metrics|Quality Control Metrics|tumour|TGZ|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WGS/qc\_metrics/tumour/TCRB-CA.DO262483.SA622771.wgs.20211117.TCRBOA7T-WGS-L3.ubam\_qc\_metrics.tgz
DNA Seq Alignment|qc\_metrics|Quality Control Metrics|tumour|TGZ|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WGS/qc\_metrics/tumour/TCRB-CA.DO262483.SA622771.wgs.20211117.TCRBOA7T-WGS-L2.ubam\_qc\_metrics.tgz
DNA Seq Alignment|qc\_metrics|Quality Control Metrics|tumour|TGZ|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WGS/qc\_metrics/tumour/TCRB-CA.DO262483.SA622771.wgs.20211117.TCRBOA7T-WGS-L1.ubam\_qc\_metrics.tgz
DNA Seq Alignment|qc\_metrics|Quality Control Metrics|normal|TGZ|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WGS/qc\_metrics/normal/TCRB-CA.DO262483.SA622770.wgs.20211118.aln.cram.qc\_metrics.tgz
DNA Seq Alignment|qc\_metrics|Quality Control Metrics|normal|TGZ|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WGS/qc\_metrics/normal/TCRB-CA.DO262483.SA622770.wgs.20211118.aln.cram.oxog\_metrics.tgz
DNA Seq Alignment|qc\_metrics|Quality Control Metrics|normal|TGZ|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WGS/qc\_metrics/normal/TCRB-CA.DO262483.SA622770.wgs.20211118.aln.cram.duplicates\_metrics.tgz
DNA Seq Alignment|qc\_metrics|Quality Control Metrics|normal|TGZ|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WGS/qc\_metrics/normal/TCRB-CA.DO262483.SA622770.wgs.20211118.TCRBOA7N-WGS-L2.ubam\_qc\_metrics.tgz
DNA Seq Alignment|qc\_metrics|Quality Control Metrics|normal|TGZ|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WGS/qc\_metrics/normal/TCRB-CA.DO262483.SA622770.wgs.20211118.TCRBOA7N-WGS-L1.ubam\_qc\_metrics.tgz
DNA Seq Alignment|qc\_metrics|Metadata|tumour|JSON|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WGS/qc\_metrics/tumour/ffd8176a-62db-4dbb-9817-6a62db2dbb97.qc\_metrics.json
DNA Seq Alignment|qc\_metrics|Metadata|normal|JSON|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WGS/qc\_metrics/normal/c0e5e3fb-6da6-4c8f-a5e3-fb6da63c8f5d.qc\_metrics.json
GATK Mutect2 Variant Calling|variant\_calling|Simple Nucleotide Variation|tumour|VCF|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WGS/variant\_calling/tumour/TCRB-CA.DO262483.SA622771.wgs.20211209.gatk-mutect2.somatic.snv.vcf.gz
GATK Mutect2 Variant Calling|variant\_calling|Simple Nucleotide Variation|tumour|VCF|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WGS/variant\_calling/tumour/TCRB-CA.DO262483.SA622771.wgs.20211209.gatk-mutect2.somatic.indel.vcf.gz
GATK Mutect2 Variant Calling|variant\_calling|Simple Nucleotide Variation|tumour|TBI|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WGS/variant\_calling/tumour/TCRB-CA.DO262483.SA622771.wgs.20211209.gatk-mutect2.somatic.snv.vcf.gz.tbi
GATK Mutect2 Variant Calling|variant\_calling|Simple Nucleotide Variation|tumour|TBI|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WGS/variant\_calling/tumour/TCRB-CA.DO262483.SA622771.wgs.20211209.gatk-mutect2.somatic.indel.vcf.gz.tbi
GATK Mutect2 Variant Calling|variant\_calling|Metadata|tumour|JSON|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WGS/variant\_calling/tumour/b97a8a3f-ad63-4215-ba8a-3fad63c215c7.variant\_calling.json
GATK Mutect2 Variant Calling|variant\_calling|Metadata|tumour|JSON|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WGS/variant\_calling/tumour/44193e58-89ad-434a-993e-5889ad434a8c.variant\_calling.json
GATK Mutect2 Variant Calling|qc\_metrics|Quality Control Metrics|tumour|TGZ|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WGS/qc\_metrics/tumour/TCRB-CA.DO262483.SA622771.wgs.20211209.gatk-mutect2.somatic.mutect\_filtering\_metrics.tgz
GATK Mutect2 Variant Calling|qc\_metrics|Quality Control Metrics|tumour|TGZ|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WGS/qc\_metrics/tumour/TCRB-CA.DO262483.SA622771.wgs.20211209.gatk-mutect2.somatic.mutect\_callable\_metrics.tgz
GATK Mutect2 Variant Calling|qc\_metrics|Quality Control Metrics|tumour|TGZ|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WGS/qc\_metrics/tumour/TCRB-CA.DO262483.SA622771.wgs.20211209.gatk-mutect2.somatic.contamination\_metrics.tgz
GATK Mutect2 Variant Calling|qc\_metrics|Quality Control Metrics|tumour|TGZ|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WGS/qc\_metrics/tumour/TCRB-CA.DO262483.SA622770.wgs.20211209.gatk-mutect2.somatic.contamination\_metrics.tgz
GATK Mutect2 Variant Calling|qc\_metrics|Metadata|tumour|JSON|https://object.cancercollaboratory.org:9080/swift/v1/genomics-public-data/benchmark-datasets/TCRB-CA/DO262483/WGS/qc\_metrics/tumour/9b5a6c13-caa5-492d-9a6c-13caa5192db1.qc\_metrics.json

